patches/server/0965-Make-banners-modifiable-with-both-BannerMeta-and-Blo.patch

```diff
@@ -0,0 +1,270 @@
+     static final ItemMetaKey COLOR = new ItemMetaKey("Color", "color");
+     static final ItemMetaKey PATTERN = new ItemMetaKey("Pattern", "pattern");
++    // Paper start - Add banner materials to CraftMetaBlockState's material list
++    static {
++        CraftMetaBlockState.BLOCK_STATE_MATERIALS.addAll(CraftMetaBanner.BANNER_MATERIALS);
++    }
++    // Paper end - Add banner materials to CraftMetaBlockState's material list
+ 
+     private DyeColor base;
+     private List<Pattern> patterns = new ArrayList<Pattern>();
+-
++    // Paper start - Create new constructor to properly handle different banners
++    private final org.bukkit.block.Banner bannerBlock;
++    @Deprecated
+     CraftMetaBanner(CraftMetaItem meta) {
+-        super(meta);
++        this(meta, Material.WHITE_BANNER);
++
++        new org.bukkit.plugin.AuthorNagException("Usage of internal deprecated constructor in paper: CraftMetaBanner(CraftMetaItem), this will incorrectly assume the item is a white banner").printStackTrace();
++    }
++
++    CraftMetaBanner(CraftMetaItem meta, Material material) {
++        super(meta, material);
+ 
+-        if (!(meta instanceof CraftMetaBanner)) {
++        if (true || !(meta instanceof CraftMetaBanner)) {
++            this.bannerBlock = (org.bukkit.block.Banner) this.getBlockState();
++            // Paper end - Create new constructor to properly handle different banners
+             return;
+         }
+ 
+@@ -74,11 +89,20 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+         this.base = banner.base;
+         this.patterns = new ArrayList<Pattern>(banner.patterns);
+     }
+-
++    // Paper start - Create new constructor to properly handle different banners
++    @Deprecated
+     CraftMetaBanner(CompoundTag tag) {
+-        super(tag);
++        this(tag, Material.WHITE_BANNER);
++
++        new org.bukkit.plugin.AuthorNagException("Usage of internal deprecated constructor in paper: CraftMetaBanner(CompoundTag), this will incorrectly assume the item is a white banner").printStackTrace();
++    }
+ 
+-        if (!tag.contains("BlockEntityTag")) {
++    CraftMetaBanner(CompoundTag tag, Material material) {
++        super(tag, material);
++
++        if (true || !tag.contains("BlockEntityTag")) {
++            this.bannerBlock = (org.bukkit.block.Banner) this.getBlockState();
++            // Paper end - Create new constructor to properly handle different banners
+             return;
+         }
+ 
+@@ -102,6 +126,12 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+ 
+     CraftMetaBanner(Map<String, Object> map) {
+         super(map);
++        // Paper start - Use underlying banner block state
++        if (true) {
++            this.bannerBlock = (org.bukkit.block.Banner) this.getBlockState();
++            return;
++        }
++        // Paper end - Use underlying banner block state
```

Comment:
Some extra stuff has to be done here to still support deserialization of this meta. You can't just ignore this constructor. You have to basically construct the blockstate with the values provided in the map. Calling the super() constructor here with that map doesn't make sense anymore because the supertype has changed to CraftMetaBlockState.